### PR TITLE
feat: also trigger subscription change events on switch

### PR DESCRIPTION
**Pull request by leogermani**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Also trigger the data events for subscription changes when there's a subscription switch.

### How to test the changes in this Pull Request:

Questions:

1. Do the docs look ok?
2. Do you think we should add an additional param to the event identifying the type of change (status or switch)? - Note that we don't have access to the previous state of the subscription to add to the event exactly what changed
3. How do I configure a non donation subscription that can be downgraded/upgraded to another product?
4. Should we keep this only to donations for now?

Testing donations:

1. As a reader, make a recurring donation
2. Go to My Account > My Subscription
3. Click on "Change price"
5. choose a product and a value and submit
6. Visit card and proceed to checkout
7. Complete checkout
8. Check that the donation_changed event is fired with the correct info (you can set a webhook using https://webhook.site/ for that if you want)

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208541920545689

### CODE REVIEW

**Review by dkoo — APPROVED** [review]
@leogermani this works as described. The "switch" functionality is related to the Name Your Price extension—to create a group of non-donation products that can be switched, you can either create a variable subscription and mark its variations as "Name Your Price" enabled, or create a bunch of products with NYP enabled and group them under a parent product (like with our Donation products).

File: includes/data-events/listeners.php [hunk 1]
```diff
@@ -252,6 +252,46 @@ function( $subscription, $status_to, $status_from ) {
...
+		}
+
+		$data = \Newspack\Data_Events\Utils::get_recurring_donation_data( $subscription );
+		if ( ! is_array( $data ) ) {
+			return;
+		}
+		return array_merge(
+			$data,
+			[
+				'status_before' => $subscription->get_status(),
```
Comment (dkoo) [inline includes/data-events/listeners.php:288]: Nit: I'm not sure we need to add `status_before` and `status_after` params to the switch handlers. These are only relevent if the status has changed

**Comment by dkoo** [comment]
To address your Qs:

> 1. Do the docs look ok?

They look fine to me—we might consider mentioning Name Your Price as the source of the "switch" feature.

> 2. Do you think we should add an additional param to the event identifying the type of change (status or switch)? - Note that we don't have access to the previous state of the subscription to add to the event exactly what changed

Not sure this is necessary if you take my [suggestion](https://github.com/Automattic/newspack-plugin/pull/3514#discussion_r1830202091) because the status changes will get `status_before` and `status_after` params. But it can't hurt to add `'subscription_switch' => true` to these handlers.

> 4. Should we keep this only to donations for now?

Nah, I think it's good to handle both. I'm not sure if any publishers use NYP for anything other than donations, but you never know.

File: includes/data-events/listeners.php (hunk 1 above)
Comment (leogermani) [reply includes/data-events/listeners.php:288]: I did that to keep consistency and make sure the event always has the same parameters in all cases. I was thinking about making it easier for people building integrations that rely on this event

**Comment by leogermani** [comment]
> @leogermani this works as described. The "switch" functionality is related to the Name Your Price extension—to create a group of non-donation products that can be switched, you can either create a variable subscription and mark its variations as "Name Your Price" enabled, or create a bunch of products with NYP enabled and group them under a parent product (like with our Donation products).

I think NYP works on top of it, but Subscription Switch is a concept that exist in Woo Subscriptions itself -> https://woocommerce.com/document/subscriptions/switching-guide/

Or am I missing something?

**Comment by dkoo** [comment]
Ah, you're right—I think the only requirement is that the product is part of a group or a variation in a variable subscription product.

File: includes/data-events/listeners.php (hunk 1 above)
Comment (dkoo) [reply includes/data-events/listeners.php:288]: That makes sense! In that case, it can't hurt to add a `'type' => 'switch' | 'status'` param too

**Comment by leogermani** [comment]
I'll merge this as it is. Thinking that adding a `type` with a value of `switch` could add more confusion instead of helping. What is a "switch"? What is "status"?... 

They can simply check all the fields and they'll know what have changed

**Pull request merged**